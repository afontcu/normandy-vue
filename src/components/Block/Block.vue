<template>
  <div class="o-block" :class="[ classes ]">

    <div class="o-block__img">
      <img v-bind="{ src, alt }" />
    </div>

    <div class="o-block__body">
      <slot></slot>
    </div>

  </div>
</template>

<script>
const VALIGN_VALUES = [ 'left', 'right' ]
const SIZE_VALUES = [ 'flush', 'tiny', 'small', 'large', 'huge' ]

export default {
  name: 'Block',
  props: {
    src: VueTypes.string.isRequired,
    alt: VueTypes.string.isRequired,
    size: VueTypes.oneOf(SIZE_VALUES),
    align: VueTypes.oneOf(ALIGN_VALUES)
  },
  data () {
    return {}
  },
  computed: {
    classes () {
      return {
        [`o-media--${this.align}`]: this.align,
        [`o-media--${this.size}`]: this.size
      }
    }
  }
}
</script>
