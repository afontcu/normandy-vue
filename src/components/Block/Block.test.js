import Block from './Block'
import { mount } from 'avoriaz'
import { compileToFunctions } from 'vue-template-compiler'

const propsData = {
  src: 'http://example.com',
  alt: 'alt'
}

describe('Block.vue', () => {
  beforeAll(() => {
    console.error = jest.fn(error => { throw new Error(error) })
    console.warn = jest.fn(warn => { throw new Error(warn) })
  })

  it('renders correct nodes when required props are provided', () => {
    const wrapper = mount(Block, { propsData })
    expect(wrapper.hasClass('o-block')).toBe(true)
    expect(wrapper.find('.o-block__body')).toBeTruthy()
    expect(wrapper.find('.o-block__img')).toBeTruthy()
  })

  it('renders slot content when defined', () => {
    const compiled = compileToFunctions('<div class="dummy"></div>')
    const wrapper = mount(Block, { propsData, slots: { default: compiled } })

    expect(wrapper.first('.dummy')).toBeTruthy()
  })

  it('src prop is required, is a String and validates correctly', () => {
    const wrapper = mount(Block, { propsData })
    const src = wrapper.vm.$options.props.src

    expect(src.required).toBeTruthy()
    expect(src.type).toBe(String)
  })

  it('alt prop is required, is a String and validates correctly', () => {
    const wrapper = mount(Block, { propsData })
    const alt = wrapper.vm.$options.props.alt

    expect(alt.required).toBeTruthy()
    expect(alt.type).toBe(String)
  })

  it('adds size class when size prop is provided', () => {
    const values = ['flush', 'tiny', 'small', 'large', 'huge']

    values.forEach((size) => {
      const wrapper = mount(Block, { propsData })
      wrapper.setProps({ size })
      expect(wrapper.hasClass(`o-block--${size}`)).toBe(true)
    })
  })

  it('adds align class when align prop is provided', () => {
    const values = ['left', 'right']

    values.forEach((align) => {
      const wrapper = mount(Block, { propsData })
      wrapper.setProps({ align })
      expect(wrapper.hasClass(`o-block--${align}`)).toBe(true)
    })
  })

  it('adds all optional classes when all props are provided', () => {
    const wrapper = mount(Block, { propsData })
    wrapper.setProps({
      size: 'small',
      align: 'left'
    })

    expect(wrapper.hasClass('o-block--small')).toBe(true)
    expect(wrapper.hasClass('o-block--left')).toBe(true)
  })

  it('changes size prop when its class', () => {
    const wrapper = mount(Block, { propsData })

    wrapper.setProps({ size: 'large' })
    expect(wrapper.hasClass('o-block--large')).toBe(true)
    expect(wrapper.hasClass('o-block--small')).toBe(false)

    wrapper.setProps({ size: 'small' })
    expect(wrapper.hasClass('o-block--small')).toBe(true)
    expect(wrapper.hasClass('o-block--large')).toBe(false)
  })

  it('breaks when required props are not provided', () => {
    expect(() => mount(Block)).toThrow()
  })

  it('breaks with invalid size prop value', () => {
    const size = 'invalidProp'
    expect(() => {
      mount(Block, { propsData: { ...propsData, size } })
    }).toThrow()
  })

  it('breaks with invalid align prop value', () => {
    const align = 'invalidProp'
    expect(() => {
      mount(Block, { propsData: { ...propsData, align } })
    }).toThrow()
  })

  it('renders an snapshot', () => {
    const wrapper = mount(Block, { propsData })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
