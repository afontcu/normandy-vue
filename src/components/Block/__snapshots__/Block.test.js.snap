// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block.vue renders an snapshot 1`] = `

<div class="o-block">
  <div class="o-block__img">
    <img src="http://example.com"
         alt="alt"
    >
  </div>
  <div class="o-block__body">
  </div>
</div>

`;
